src/hnsw/product_quantization.c

```diff
@@ -0,0 +1,275 @@
+
+    for(i = 0; i < point_cnt; i++) {
+        for(j = 0; j < subset_dim; j++) {
+            if(!mean[ j ]) mean[ j ] = 0;
+            mean[ j ] += points[ i ][ j ];
+        }
+    }
+
+    for(i = 0; i < subset_dim; i++) {
+        mean[ i ] = mean[ i ] / point_cnt;
+    }
+
+    return mean;
+}
+
+/*
+ * Update cluster centers with the mean values of assigned points
+ * And remove points from cluster
+ * */
+static void update_centers(Cluster **clusters, uint32 k, uint32 subset_dim)
+{
+    uint32   i;
+    Cluster *current_cluster;
+
+    for(i = 0; i < k; i++) {
+        current_cluster = clusters[ i ];
+        if(!current_cluster->point_cnt) continue;
+
+        current_cluster->center = calculate_mean(current_cluster->points, current_cluster->point_cnt, subset_dim);
+
+        current_cluster->points = NULL;
+        current_cluster->point_cnt = 0;
+    }
+}
+
+/*
+ * Checks whether distances between new and old centers
+ * are less then the defined threshold to stop iterations
+ * and return current centers as result.
+ * */
+bool should_stop_iterations(float4              **old_centers,
+                            Cluster             **clusters,
+                            uint32                cluster_count,
+                            uint32                subset_dim,
+                            usearch_metric_kind_t distance_metric)
+{
+    usearch_error_t error = NULL;
+    uint32          i;
+    float4          threshold = 0.15f;
+    float4          distance = 0.0f;
+
+    for(i = 0; i < cluster_count; i++) {
+        distance += usearch_distance(
+            old_centers[ i ], clusters[ i ]->center, usearch_scalar_f32_k, subset_dim, distance_metric, &error);
+        assert(!error);
+    }
+
+    distance /= cluster_count;
+
+    return distance <= threshold;
+}
+
+/*
+ * Run k-means clustering over the dataset in a subset of vectors
+ * args:
+ *   k - cluster count
+ *   dataset - 2d array of float numbers
+ *   dataset_size - length of dataset (axis 0)
+ *   subset_start - index from which the subset will be taken (in case if subset count is 1 this will be 0)
+ *   subset_dim - length of the slice to take from dataset vectors starting from subset_start index
+ *   iter - number of iterations to run the algorithm
+ * returns:
+ *   2D array of floats. [ [f32] x k ] each sub array will be a centroid point
+ */
+float4 **k_means(uint32                k,
+                 float4              **dataset,
+                 uint32                dataset_size,
+                 uint32                subset_start,
+                 uint32                subset_dim,
+                 usearch_metric_kind_t distance_metric,
+                 uint32                iter)
+{
+    uint32    i, j;
+    float4  **centroids = palloc(sizeof(size_t) * k);
+    Cluster **clusters = initialize_clusters(k, dataset, dataset_size, subset_start);
+    float4  **old_centers = palloc(sizeof(size_t) * k);
+
+    for(i = 0; i < iter; i++) {
```

Comment:
if it is safe to do so, once in every iteration call CHECK_FOR_INTERRUPTS() to make sure the client is able to cancel the request if they change their mind by sending a signal to the process (e.g. via CTRL-C)